### PR TITLE
Enable async support

### PR DESCRIPTION
**Pull request by bmwant**
## Rationale
PR enabling async support #152 

## Open questions
@asvetlov @jettify see comments below within code

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -4,11 +4,8 @@ CHANGES
 0.15.0 (2017-xx-xx)
 -------------------
 
--
+- Enable async support feature available in Jinja 2.9 #152
```
Comment (bmwant) [inline CHANGES.txt]: do we update this file manually referencing issue number (e.g. 152 here)?

File: aiohttp_jinja2/__init__.py [hunk 2]
```diff
@@ -18,8 +18,8 @@
 
 
 def setup(app, *args, app_key=APP_KEY, context_processors=(),
-          filters=None, default_helpers=True, **kwargs):
-    env = jinja2.Environment(*args, **kwargs)
+          filters=None, default_helpers=True, enable_async=True, **kwargs):
+    env = jinja2.Environment(*args, enable_async=enable_async, **kwargs)
```
Comment (bmwant) [inline aiohttp_jinja2/__init__.py]: should I make this keyword only argument?

File: setup.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 def read(f):
     return open(os.path.join(os.path.dirname(__file__), f)).read().strip()
 
-install_requires = ['aiohttp>=0.20', 'jinja2>=2.7']
+install_requires = ['aiohttp>=0.20', 'jinja2>=2.9']
```
Comment (bmwant) [inline setup.py]: Do we need to bump version here or should I add an explicit check within setup method that will enable the flag only in case Jinja version is higher than `2.9`. First option means that we kinda break backwards compatibility for newest release of `aiohttp_jinja2` with older `jinja2` versions

**Comment by samuelcolvin** [comment]
LGTM, I think you can safely uprev. jinja to 2.9. I very much doubt anyone will have jinja pinned to <2.9.

I think it would be good to add an examples to the docs of how async can be used.

File: CHANGES.txt (hunk 1 above)
Comment (jettify) [reply CHANGES.txt]: yes manually, please share if you know better way.

File: aiohttp_jinja2/__init__.py (hunk 2 above)
Comment (jettify) [reply aiohttp_jinja2/__init__.py]: `enable_async` disabled by default?

Comment (samuelcolvin) [reply aiohttp_jinja2/__init__.py]: surely we should check and find out how big the performance impact is? If it's negligible then it can be enabled by default, else not.

File: setup.py (hunk 3 above)
Comment (jettify) [reply setup.py]: I think it should be fine, for older versions of jinja2 one can use older version of aiohttpp-jinja2

**Comment by jettify** [comment]
Agree with @samuelcolvin lets add, example with `enable_async=True`

File: aiohttp_jinja2/__init__.py (hunk 2 above)
Comment (samuelcolvin) [reply aiohttp_jinja2/__init__.py]: I'll try running aio-libs/aiohttp-benchmarks with and without async on the fortunes test and see what happens.

**Comment by bmwant** [comment]
@samuelcolvin @jettify there is no difference in high-level api for the end-user to work within this mode, so no point to add some examples. anyway, I've updated readme section mentioning that.

**Comment by samuelcolvin** [comment]
> there is no difference in high-level api for the end-user to work within this mode

I don't think that's correct, for example the following gives:


```python
import asyncio
from jinja2 import Template

template_str = 'foobar: {{ foobar() }}'

async def foobar():
    await asyncio.sleep(0.1)
    return '123'

async def with_async():
    template = Template(template_str, enable_async=True)
    v = await template.render_async(foobar=foobar)
    debug(v)

async def without_async():
    template = Template(template_str, enable_async=False)
    v = template.render(foobar=foobar)
    debug(v)

loop = asyncio.get_event_loop()
loop.run_until_complete(with_async())
loop.run_until_complete(without_async())
```

gives:
```
async_jinja.py:15 with_async
  v: 'foobar: 123' (str) len=11
<template>:12: RuntimeWarning: coroutine 'foobar' was never awaited
async_jinja.py:21 without_async
  v: 'foobar: <coroutine object foobar at 0x7f76b8294780>' (str) len=51
```

`foobar()` can only be used with `enable_async=True` so I think there is an impact on users which we should describe to users.

**Comment by bmwant** [comment]
@samuelcolvin I see what you mean. I was talking about using only `render_string`, `render_template` and `template` decorators from `aiohttp-jinja2` itself

**Comment by samuelcolvin** [comment]
### Another question:

Should we support streamed responses using [`async_generate()`](http://jinja.pocoo.org/docs/2.9/api/#jinja2.Template.generate_async)? I think this could be very cool, particularly for large pages where users could start interacting with buttons at the top of the page while the database is still churning away doing queries to render the bottom of the page.

For example, in the following I can click the button at the top while the rest of the html is still loading:

```python
import asyncio

from aiohttp import web
from jinja2 import Template


template_str = """
<h3>Top menu</h3>
<button>Top button</button>

<h3>Complex data</h3>
<ul>
  {% for v in get_data() -%}
    <li>{{ v }}</li>
  {% endfor -%}
</ul>
"""


async def get_data():
    for i in range(10):
        await asyncio.sleep(0.5)
        yield i


async def handle(request):
    template = Template(template_str, enable_async=True)

    r = web.StreamResponse()
    r.content_type = 'text/html'
    await r.prepare(request)
    async for chunk in template.generate_async(get_data=get_data):
        r.write(chunk.encode())
    return r

app = web.Application()
app.router.add_get('/', handle)

web.run_app(app)
```

**Comment by bmwant** [comment]
I guess we can merge this one and address `StreamedResponse` features in another PR.
 @samuelcolvin @asvetlov what do you think?

**Comment by asvetlov** [comment]
I need time to learn async implementation in jinja2 first.
I didn't dig into code yet but just created an issue.

**Comment by samuelcolvin** [comment]
Life tells us that most people don't RTFM.

We should therefore do one of two things:
1. implement `enable_async=True` by default and also implement `StreamedResponse` by default so aiohttp-jinja "just works" with asynchronous rendering.
2. add `enable_async` but with `False` as default so the behaviour (and performance) doesn't change unless people explicitly enable it.

**Comment by jettify** [comment]
@samuelcolvin have you managed to benchmark async/mode?

**Comment by jettify** [comment]
As for options lets do simplest one: make `enable_async=False` in initial release and upgrade after.

**Comment by samuelcolvin** [comment]
Ok, I've run some simple performance testing, results [here](https://gist.github.com/samuelcolvin/8362662a9c3aade012bc2e107024287c), summary:

```
Requests/sec:
    master:                               596.69
    enable_async=False                    594.52
    enable_async=True                     219.19
    enable_async=True + StreamResponse    9.89
```

Note the `StreamResponse` case I had to hack the package to work with python 3.6 then add `generate_async` support. It worked fine when simply reloading the page a few times in chrome but when load testing it with wrk I got `socket.send() raised exception.` printed in about 10k times.

In summary I agree with @jettify; we should release with `enable_async=False` and think carefully about how to work properly with async/render_async/generate_async.

**Comment by dpallot** [comment]
Any idea when you guys are going to merge and release this?

**Comment by asvetlov** [comment]
After digging into jinja2 internals I've found that the idea is not very good actually.
It hides async calls in templates, explicit support for `await` and `async for` in template language should be much better. I'd like to see explicit yield points, Guido van Rossum emphasizes this syntax as one of most important features in asyncio world.
Other drawback is performance: jinja2 checks every function by `isawaitable()`, it  slows down the whole engine regardless async functions was actually called or not.

**Comment by dpallot** [comment]
@asvetlov That's a shame. I will keep my template rendering in a thread pool for the time being.

**Comment by hellysmile** [comment]
We do same thing with treadpool, maybe it makes sense to add thredpool-based rendering to aiohttp-jinja?

**Comment by asvetlov** [comment]
#156

**Comment by ccxtechnologies** [comment]
Is this going to be merged in as an optional feature disabled by default? I want to await on coroutines from my jinja scripts, I'm not too concerned with performance implications, this works great for that.

Maybe someday the jinja folks will make coroutine awaits in their scripts explicitly defined (which would make way more sense to me too) but for now this works.

**Comment by asvetlov** [comment]
Not sure if I want to have the feature merged.
It encourages bad style: calling async functions without `await` keyword makes a mess.

**Comment by ccxtechnologies** [comment]
No worries, I understand. At fist I had tried adding awaits to my scripts, it took me a while to figure out they weren't supposed to be there.

I'll just create a fork off of this PR so I don't lose it if you decide not to include it. I need this feature for my project.

**Comment by bmwant** [comment]
@asvetlov am I get this right: we do not want this feature to be in for now? I'm closing PR and corresponding feature then.

**Comment by asvetlov** [comment]
I pretty sure jinja should add native support for `await` and `async with`/`async for`.
Unfortunately I have no time to make a fix for jinja myself
